### PR TITLE
Fix optimize read in order from a view.

### PR DESCRIPTION
**Pull request by vitlibar**
### Changelog category:
- Performance Improvement

### Changelog entry:
Enable ORDER BY optimization for reading data in corresponding order from a MergeTree table in case that the table is behind a view.

For example, the following SELECT

```
CREATE TABLE table1 (x UInt64) ENGINE=MergeTree ORDER BY x;
CREATE VIEW view1 AS SELECT x FROM table1;
...
SELECT * FROM (SELECT * FROM view1) ORDER BY x DESC LIMIT 20;
```

must understand that the table is already ordered and read only a few rows from the table (before this PR this query used to read the full table).

### CODE REVIEW

File: src/Functions/materialize.h [hunk 1]
```diff
@@ -52,6 +52,15 @@ class FunctionMaterialize : public IFunction
...
     }
+
+    bool hasInformationAboutMonotonicity() const override { return true; }
+
+    Monotonicity getMonotonicityForRange(const IDataType &, const Field &, const Field &) const override
+    {
+        /// Depending on the argument the function materialize() is either a constant or works as identity().
+        /// In both cases this function is monotonic and non-decreasing.
+        return {.is_monotonic = true, .is_always_monotonic = true};
+    }
```
Comment (vitlibar) [inline src/Functions/materialize.h:63]: `StorageView` implicitly applies the `materialize()` function to each column of the result of an internal query (see [StorageView::read](https://github.com/ClickHouse/ClickHouse/blob/eeb2c5b223730af0f3d410cb7a780ec859d35f37/src/Storages/StorageView.cpp#L169C19-L170C1)). And then [optimize_read_in_order](https://github.com/ClickHouse/ClickHouse/blob/eeb2c5b223730af0f3d410cb7a780ec859d35f37/src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp#L341) checks the monotonicity of applied functions (see [matchTrees()](https://github.com/ClickHouse/ClickHouse/blob/eeb2c5b223730af0f3d410cb7a780ec859d35f37/src/Processors/QueryPlan/Optimizations/actionsDAGUtils.h#L13)). So if `materialize()` becomes monotonic then the optimization `optimize_read_in_order` can work for views.

**Comment by vitlibar** [comment]
Also `EXPLAIN` shows the change this PR makes:

Before:
```
> EXPLAIN actions=1 SELECT * FROM (SELECT * FROM view1) ORDER BY x DESC LIMIT 20;

┌─explain─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ Expression (Projection)                                                                                                                                                                                         │
│ Actions: INPUT :: 0 -> x UInt64 : 0                                                                                                                                                                             │
│ Positions: 0                                                                                                                                                                                                    │
│   Limit (preliminary LIMIT (without OFFSET))                                                                                                                                                                    │
│   Limit 20                                                                                                                                                                                                      │
│   Offset 0                                                                                                                                                                                                      │
│     Sorting (Sorting for ORDER BY)                                                                                                                                                                              │
│     Sort description: x DESC                                                                                                                                                                                    │
│     Limit 20                                                                                                                                                                                                    │
│       Expression ((Before ORDER BY + (Projection + (Before ORDER BY + (Convert VIEW subquery result to VIEW table structure + (Materialize constants after VIEW subquery + (Projection + Before ORDER BY))))))) │
│       Actions: INPUT : 0 -> x UInt64 : 0                                                                                                                                                                        │
│                FUNCTION materialize(x :: 0) -> materialize(x) UInt64 : 1                                                                                                                                        │
│                ALIAS materialize(x) :: 1 -> x UInt64 : 0                                                                                                                                                        │
│       Positions: 0                                                                                                                                                                                              │
│         ReadFromMergeTree (default.table1)                                                                                                                                                                      │
│         ReadType: Default                                                                                                                                                                                       │
│         Parts: 1                                                                                                                                                                                                │
│         Granules: 1                                                                                                                                                                                             │
└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘

```

After:
```
> EXPLAIN actions=1 SELECT * FROM (SELECT * FROM view1) ORDER BY x DESC LIMIT 20;

┌─explain─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ Expression (Projection)                                                                                                                                                                                         │
│ Actions: INPUT :: 0 -> x UInt64 : 0                                                                                                                                                                             │
│ Positions: 0                                                                                                                                                                                                    │
│   Limit (preliminary LIMIT (without OFFSET))                                                                                                                                                                    │
│   Limit 20                                                                                                                                                                                                      │
│   Offset 0                                                                                                                                                                                                      │
│     Sorting (Sorting for ORDER BY)                                                                                                                                                                              │
│     Prefix sort description: x DESC                                                                                                                                                                             │
│     Result sort description: x DESC                                                                                                                                                                             │
│     Limit 20                                                                                                                                                                                                    │
│       Expression ((Before ORDER BY + (Projection + (Before ORDER BY + (Convert VIEW subquery result to VIEW table structure + (Materialize constants after VIEW subquery + (Projection + Before ORDER BY))))))) │
│       Actions: INPUT : 0 -> x UInt64 : 0                                                                                                                                                                        │
│                FUNCTION materialize(x :: 0) -> materialize(x) UInt64 : 1                                                                                                                                        │
│                ALIAS materialize(x) :: 1 -> x UInt64 : 0                                                                                                                                                        │
│       Positions: 0                                                                                                                                                                                              │
│         ReadFromMergeTree (default.table1)                                                                                                                                                                      │
│         ReadType: InReverseOrder                                                                                                                                                                                │
│         Parts: 1                                                                                                                                                                                                │
│         Granules: 1                                                                                                                                                                                             │
└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘

```

**Comment by SaltTan** [comment]
related #34720
maybe also #18561?

File: tests/queries/0_stateless/02877_optimize_read_in_order_from_view.sql [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+CREATE VIEW view1 AS SELECT number FROM table1;
+
+-- The following SELECT is expected to read 20 rows. In fact it may decide to read more than 20 rows, but not too many anyway.
+-- So we'll check that the number of read rows is less than 40.
+
+SELECT /* test 02877, query 1 */ * FROM (SELECT * FROM view1) ORDER BY number DESC LIMIT 20 SETTINGS log_queries=1;
+
+SYSTEM FLUSH LOGS;
+
+SELECT concat('read_rows=', if(read_rows<40, 'ok', toString(read_rows))) FROM system.query_log WHERE current_database = currentDatabase() AND query LIKE '%test 02877, query 1%' AND type='QueryFinish';
```
Comment (nickitat) [inline tests/queries/0_stateless/02877_optimize_read_in_order_from_view.sql:18]: imo it is better to check the query plan

Comment (vitlibar) [reply tests/queries/0_stateless/02877_optimize_read_in_order_from_view.sql]: Ok, I've rewritten the test.

**Comment by vitlibar** [comment]
> related #34720 maybe also #18561?

https://github.com/ClickHouse/ClickHouse/issues/34720 - yes
https://github.com/ClickHouse/ClickHouse/issues/18561 - no

**Review by nickitat — APPROVED** [review]

**Pull request merged**